### PR TITLE
Add Bazel config files and automated tests

### PR DESCRIPTION
**Pull request by 95-martin-orion**
Things to try:
```
# Run all unit tests
bazel test tests:all

# Run a sample simulation (takes ~15 seconds)
bazel run apps:qsim_base -- -c circuits/circuit_q24
```

Remaining work:
- Document Bazel support
- Test with Mac / Windows

### CODE REVIEW

File: .github/workflows/bazeltest.yml [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+  pull_request:
+
+jobs:
+  # Run tests with Bazel.
+  test:
+    name: Test on ${{ matrix.os }}
+    runs-on: ${{ matrix.os }}
+    strategy:
+      matrix:
+        os: [ubuntu-latest]
```
[the remark this answers is not in the document]
Comment (95-martin-orion) [reply .github/workflows/bazeltest.yml]: Done.

[the remark this answers is not in the document]
Comment (95-martin-orion) [reply .github/workflows/bazeltest.yml]: Looks like the older Bazel version is disagreeing with the `googletest` submodule. I'll take a closer look.

**Comment by 95-martin-orion** [comment]
> I would say thematically there are a few things you might need to modify: {...}

Took a shot at (1) and (2), though there's work required elsewhere to properly support it. Testing with `--config=avx` works on my machine, but it looks like the older versions aren't quite there yet.

**Comment by 95-martin-orion** [comment]
Updated `WORKSPACE` shamelessly replicated from TFQ.

The current version passes on my machine with `bazel==0.26.0`, which suggests that the latest failure stems from using `ubuntu-16.04`. I have a lead for this, but it may take a bit as the implementation looks fairly invasive.

**Comment by 95-martin-orion** [comment]
Possible cause of failure on `ubuntu-16.04`:
- Original `ubuntu-16.04` release was [April 2016](https://wiki.ubuntu.com/Releases)
- _mm256_cvtss_f32 was added in [April 2017](https://github.com/gcc-mirror/gcc/commit/dcb2c5271869ed6829ca3469946a8ae90a1c1cdb)

**Comment by 95-martin-orion** [comment]
> Possible cause of failure on `ubuntu-16.04`: {...}

Replacing the operator with its implementation seems to resolve this; unfortunately, this fix might not work across all systems.

**Comment by MichaelBroughton** [comment]
I think you can downgrade this instruction pretty easily. Maybe something like this:
```C
float buffer[8];
_mm256_storeu_ps(buffer, s1);
return (float)(buffer[0] + buffer[1] + buffer[2] + buffer[3] .. + ...);
```
This is 32bit float and is squaring the number and adding them up over the whole state. I think this might be prone to underflow... can you double check that `norm` doesn't underflow on states with lots of small amplitudes ?

**Comment by 95-martin-orion** [comment]
> Can you double check that `norm` doesn't underflow on states with lots of small amplitudes ?

Thankfully, this is a non-issue. `Norm` is only used in the simulator unit tests, and it is always expected to output `1`. Even if underflow occurred, it would be too small to affect this result.

**Comment by 95-martin-orion** [comment]
@sergeisakov for approval on qsim changes.

> Now on to msan asan and multiple platforms in future PRs :) ?

Onward!

**Pull request merged**